### PR TITLE
Z3 term implementation

### PR DESCRIPTION
**Pull request by lstuntz**
Implementation of Z3 terms. Note, termIter, testing, and datatype terms are not included in this pr

### CODE REVIEW

**Comment by lstuntz** [comment]
For this pr ignore the test file and formatting of all files -- I'm going to do another pr after this with formatting and with separating out testing.

**Comment by amaleewilson** [comment]
@makaim we can't seem to add you to this PR

**Comment by amaleewilson** [comment]
@makaimann we can't add you as a reviewer to this PR? 

Can you review?

File: tests/z3/z3_test.cpp [hunk 1]
```diff
@@ -6,36 +6,169 @@
 
 #include "z3_factory.h"
 
+#include "z3_factory.h"
```
Comment (makaimann) [inline tests/z3/z3_test.cpp]: minor: duplicated include -- probably from resolving merge conflicts.

File: tests/z3/z3_test.cpp [hunk 2]
```diff
@@ -6,36 +6,169 @@
...
+
 using namespace smt;
 using namespace std;
+//using namespace z3;
+
+void printVec(string name, TermVec ts) {
```
Comment (makaimann) [inline tests/z3/z3_test.cpp]: nitpicky and not that big of a deal, but could you call it `print_vec`? Just trying to use the underscore syntax for all functions even if it's in a test.

File: z3/include/z3_term.h [hunk 3]
```diff
@@ -29,60 +29,73 @@ namespace smt {
...
+public:
+	// Non-functions
+   Z3Term(expr t, context &c) : term(t), is_function(false), z_func(c) {
+	   ctx = &c;
+   };
+   // Functions
+   Z3Term(func_decl zfunc, context &c) : term(c), is_function(true), z_func(zfunc) {
+	   ctx = &c;
+   };
+	~Z3Term() {
```
Comment (makaimann) [inline z3/include/z3_term.h:67]: Looks like there's some formatting weirdness here. Maybe from the merge conflicts? Although that isn't supposed to happen. As we discussed earlier though, you can just run clang-format later, so nothing to be done now.

File: z3/src/z3_solver.cpp [hunk 4]
```diff
@@ -71,19 +130,64 @@ Term Z3Solver::get_selector(const Sort &s, std::string con,
...
 
 Term Z3Solver::make_term(const Term &val, const Sort &sort) const {
-	throw NotImplementedException(
-			"Constant arrays not supported for Z3 backend.");
+
+	std::shared_ptr<Z3Term> zterm = std::static_pointer_cast < Z3Term > (val);
+	std::shared_ptr<Z3Sort> zsort = std::static_pointer_cast < Z3Sort > (sort);
+
+	if (zsort->is_function || zterm->is_function) {
+		throw IncorrectUsageException("Arrays in functions are not supported for Z3"); //is this true?
```
Comment (makaimann) [inline z3/src/z3_solver.cpp]: Regarding the comment - that seems reasonable to me. I don't think any solver supports something like that. This function is called `make_term` so it's consistent with the rest of the functions, but is really just for creating constant arrays. E.g. an array where every element is set to a constant value like 0. Doesn't really make sense to combine it with functions (or that would at least be higher order logic). You could make the error message a bit more specific if you want. Something like: `Cannot create constant array with function element`.

File: z3/src/z3_solver.cpp [hunk 5]
```diff
@@ -252,33 +353,303 @@ Sort Z3Solver::make_sort(const Sort &sort_con, const SortVec &sorts) const {
...
+	} else {
+		// nb this is creating an expr
+		expr z_term = ctx.constant(z_name, zsort->type);
+
+		return std::make_shared<Z3Term> (z_term, ctx);
+	}
 }
 
 Term Z3Solver::make_param(const std::string name, const Sort &sort) {
 	throw NotImplementedException("make_param not supported by Z3 yet.");
```
Comment (makaimann) [inline z3/src/z3_solver.cpp]: FYI, forgot to delete the exception.

File: z3/src/z3_solver.cpp [hunk 6]
```diff
@@ -252,33 +353,303 @@ Sort Z3Solver::make_sort(const Sort &sort_con, const SortVec &sorts) const {
...
-	throw NotImplementedException(
-			"Term iteration not implemented for Z3 backend.");
+	shared_ptr<Z3Term> zterm0 = static_pointer_cast < Z3Term > (t0);
+	shared_ptr<Z3Term> zterm1 = static_pointer_cast < Z3Term > (t1);
+	shared_ptr<Z3Term> zterm2 = static_pointer_cast < Z3Term > (t2);
+	Z3_ast res;
+
+	if (!op.num_idx) {
+		if (zterm0->is_function || zterm1->is_function || zterm2->is_function){
+			throw IncorrectUsageException("Cannot make a ternary op term with a function.");
```
Comment (makaimann) [inline z3/src/z3_solver.cpp]: Just FYI, you actually could do this. For example: `make_term(Apply, f, x, y)`. However, this isn't important to fix now.  I wouldn't be surprised if other solver backends don't handle that case either. Instead they just assume that function applications always use the `make_term(Op, TermVec)` signature.

File: z3/src/z3_term.cpp [hunk 7]
```diff
@@ -63,87 +63,112 @@ bool Z3TermIter::equal(const TermIterBase & other) const
...
+		return z_func.hash();
+	}
+	return term.hash();
 }
 
-bool Z3Term::compare(const Term & absterm) const
+std::size_t Z3Term::get_id() const
 {
-throw NotImplementedException(
+  throw NotImplementedException(
```
Comment (makaimann) [inline z3/src/z3_term.cpp]: Message wasn't changed from copy-paste. Also, if you feel like it, I think Z3 does have a `get_id` method. But it's not that vital.

**Review by makaimann — APPROVED** [review]
Looks great! Just left a few comments for you to consider. And as we discussed offline, it's not clang-formatted but that can be done later before the main PR to smt-switch. Just flagging that so you/we remember to do it later.

**Pull request merged**